### ISSUE TITLE
Fix Backup and Restore in 1.1.X with Custom Certs

### ISSUE THREAD

**Issue by leflambeur** [rancher/rke#2313]
**RKE version:**
1.1.X
**Steps to Reproduce:**

RKE creation commands are as follows (adapted from ansible)

rke up --config rancher-cluster.yml --custom-certs --cert-dir rancher-cluster_certs

and generate csr with:

rke cert generate-csr --config rancher-cluster.yml

(Customer uses the full path from root as paths to files + directories in their environment). 

I can supply the ansible as well if it's helpful. To do the actual signing though they're just using the out the box ansible module https://docs.ansible.com/ansible/latest/collections/community/crypto/x509_certificate_module.html

Key used is just 2048

---

In 1.1 there is no support for restoring clusters with custom-certs

Support for this was added with the following commit to master 

https://github.com/rancher/rke/pull/1957

But only some lines were cherry-picked back to 1.1 release in https://github.com/rancher/rke/pull/2191

The following PR is raised to add the lines we've found are necessary to make sure things are parsed correctly

https://github.com/rancher/rke/pull/2307

The behaviour works in 1.2.1 as expected as it has the full fix
[succesful_restore_1_2_1.txt](https://github.com/rancher/rke/files/5466423/succesful_restore_1_2_1.txt)

gz#12779

1.1.6 which was introduced after the backport does not because it omits the relevant lines
[failed_restore_1_1_6.txt](https://github.com/rancher/rke/files/5466386/failed_restore_1_1_6.txt)

As described both environments are built exactly the same using ansible and as a pipeline - only difference is rke version

**Comment by superseb** [comment]
Can be tested using `v1.1.13-rc1`

**Comment by sowmyav27** [comment]
Reopening for an RKE RC from `release/v1.1` branch

**Comment by sadiapoddar** [comment]
Tried to reproduce the issue with the following rke versions 
 v1.1.12, v1.1.13, and v1.1.14-rc1 but couldn't reproduce. 

Steps to repro
1. Created 3 ec2 instances - 
2. Created cluster.yml file speck 1 etcd, 1 control and 1 worker node
3. Generated the custom certs 

>rke cert generate-csr --cert-dir <custom-dir-path> 

4. Followed this wiki to generate certs using CSRs and Custom CA 
https://github.com/rancherlabs/the-ranch/wiki/Testing-Custom-Certificates-and-CSR-Generation

5. Brought up the rke cluster
>rke up --custom-certs --cert-dir=custom_certs

5. Once the rke cluster was up and running took a base snapshot

 ./rke etcd snapshot-save --name snapshot1

6. Added some new namespaces and workload and took another snapshot
>./rke etcd snapshot-save --name snapshot2

7. Restore the cluster from the base snapshot1 and verified there are no custom namespaces and resources
 ./rke etcd snapshot-restore --name snapshot1 --custom-certs --cert-dir custom_certs --config cluster.yml
8. Restore the cluster from snaphsot2 and verified all the custom namespaces and resources are available in the cluster. 

9 ./rke etcd snapshot-restore --name snapshot2 --custom-certs --cert-dir custom_certs --config cluster.yml


Repeated the above steps with custom dir's full path and relative path and didn't hit any error during restoring from etcd's snapshots.

**Comment by sadiapoddar** [comment]
@leflambeur 
Hello William,
Can you please verify the steps I tried to repro this issue and let's know if we are missing any steps to repro the issue?
Thanks,
-Sadia

**Comment by sadiapoddar** [comment]
Tried to reproduce with rke 1.1.6 Steps to repro

1. Created 3 ec2 instances -
2. Created cluster.yml file spec 3 nodes with all roles(etcd, control, worker) enabled
3. Generated the custom certs

   >rke cert generate-csr --cert-dir

4. **Followed this wiki to generate certs using CSRs and Custom CA**
   https://github.com/rancherlabs/the-ranch/wiki/Testing-Custom-Certificates-and-CSR-Generation

**5. Brought up the rke cluster**

 rke up --custom-certs --cert-dir=custom_certs
 rke-test-2313-6 % ./rke up --custom-certs --cert-dir=custom_dir  
INFO[0000] Running RKE version: v1.1.6    
..........
.........
INFO[0169] [sync] Syncing nodes Labels and Taints       
INFO[0169] [sync] Successfully synced nodes Labels and Taints 
INFO[0169] [network] Setting up network plugin: canal   
INFO[0169] [addons] Saving ConfigMap for addon rke-network-plugin to Kubernetes 
INFO[0170] [addons] Successfully saved ConfigMap for addon rke-network-plugin to Kubernetes 
INFO[0170] [addons] Executing deploy job rke-network-plugin 
INFO[0180] [addons] Setting up coredns                  
INFO[0180] [addons] Saving ConfigMap for addon rke-coredns-addon to Kubernetes 
INFO[0181] [addons] Successfully saved ConfigMap for addon rke-coredns-addon to Kubernetes 
INFO[0181] [addons] Executing deploy job rke-coredns-addon 
INFO[0186] [addons] CoreDNS deployed successfully       
INFO[0186] [dns] DNS provider coredns deployed successfully 
INFO[0186] [addons] Setting up Metrics Server           
INFO[0186] [addons] Saving ConfigMap for addon rke-metrics-addon to Kubernetes 
INFO[0186] [addons] Successfully saved ConfigMap for addon rke-metrics-addon to Kubernetes 
INFO[0186] [addons] Executing deploy job rke-metrics-addon 
INFO[0192] [addons] Metrics Server deployed successfully 
INFO[0192] [ingress] Setting up nginx ingress controller 
INFO[0192] [addons] Saving ConfigMap for addon rke-ingress-controller to Kubernetes 
INFO[0192] [addons] Successfully saved ConfigMap for addon rke-ingress-controller to Kubernetes 
INFO[0192] [addons] Executing deploy job rke-ingress-controller 
INFO[0197] [ingress] ingress controller nginx deployed successfully 
INFO[0197] [addons] Setting up user addons              
INFO[0197] [addons] no user addons defined              
INFO[0197] Finished building Kubernetes cluster successfully 


**6. Took a snapshot** 
 rke-test-2313-6 % ./rke etcd  snapshot-save --name snapshot2
INFO[0000] Running RKE version: v1.1.6                  
INFO[0000] Starting saving snapshot on etcd hosts   
....
...
NFO[0016] Finished saving/uploading snapshot [snapshot2] on all etcd hosts 


Restore the cluster from the base snapshot1 and verified there are no custom namespaces and resources

**7. Restored etcd from snapshot**
 rke-test-2313-6 % ./rke etcd snapshot-restore --name snapshot2 --custom-certs --cert-dir  /Users/sadiapodar/rkestuff/rke-test-2313-6/custom_dir --config cluster.yml
INFO[0000] Running RKE version: v1.1.6                  
INFO[0000] Checking if state file is included in snapshot file for [snapshot2] 
....
....
INFO[0140] Finished building Kubernetes cluster successfully 
INFO[0140] Restarting network, ingress, and metrics pods 
INFO[0142] Finished restoring snapshot [snapshot2] on all etcd hosts 


Repeated the above restore command with custom dir's full path and relative path and didn't hit any error during restoring from etcd's snapshots with custom certs.

**Comment by jiaqiluo** [comment]
The bug is NOT reproduced with RKE CLI version `v1.1.6`

The cluster has 3 all-role nodes, and I used custom cert with 3-tier certs. 
The chain of the certs are:
```
root.pem -> kube-ca.pem -> the rest of kube-ca-*.pem
root.pem -> kube-apiserver-requestheader-ca.pem -> kube-apiserver-proxy-client.pem
```

Instead of the error that is reported in the issue when restore the etcd backup, it works on my setup and here is the output:
```
...
INFO[0053] Initiating Kubernetes cluster
INFO[0053] [certificates] GenerateServingCertificate is disabled, checking if there are unused kubelet certificates
INFO[0053] [certificates] Generating admin certificates and kubeconfig
INFO[0053] Successfully Deployed state file at [./rancher-cluster.rkestate]
INFO[0053] Building Kubernetes cluster
...
```

-----------
The bug is NOT reproduced with RKE CLI  version `v1.1.12`

The cluster has 1 all-role node, and I used custom cert with 3-tier certs (just like the above)

**Comment by jiaqiluo** [comment]
The bug is NOT reproduced with RKE CLI version v1.1.14-rc3

The cluster has 1 all-role node, and I used custom cert with 3-tier certs (just like the above)

**Comment by bmdepesa** [comment]
Closing as we cannot reproduce and this should be fixed in RKE v1.1.14. We can reopen if necessary.
